service/storage_service.cc

```diff
@@ -1536,6 +1536,39 @@ future<> storage_service::update_topology_with_local_metadata(raft::server& raft
     co_await _sys_ks.local().set_must_synchronize_topology(false);
 }
 
+// Wait until gossiper observes the given nodes or timeout is reached.
+// Called on initial startup (bootstrap/replace), not during restarts.
+static future<> wait_until_seen_during_boot(
+        const gms::gossiper& g, const std::unordered_set<gms::inet_address>& initial_contact_nodes,
+        db::timeout_clock::time_point timeout, abort_source& as) {
+    logger::rate_limit rate_limit{std::chrono::seconds{5}};
+    while (!as.abort_requested()) {
+        std::vector<gms::inet_address> observed = g.get_endpoints();
+        std::unordered_set<gms::inet_address> observed_set{observed.begin(), observed.end()};
+
+        std::vector<gms::inet_address> diff;
+        diff.reserve(initial_contact_nodes.size());
+        for (auto& n: initial_contact_nodes) {
+            if (!observed_set.contains(n)) {
+                diff.push_back(n);
+            }
+        }
+        if (diff.empty()) {
+            slogger.info("All initial contact nodes ({}) seen in gossip", initial_contact_nodes);
+            co_return;
+        }
+
+        if (db::timeout_clock::now() >= timeout) {
+            auto err = ::format("Timed out waiting for nodes {} to show up in gossip during boot", diff);
+            slogger.error("{}", err);
```

Comment:
I'm happy with this error message and exception - I think it was a mistake in existing code to just allow the timeout to be reached and not fail the boot, even though we know that we can't proceed this way.

On the other hand, I'm not sure you should have really insisted that ALL seeds be available. Isn't the whole point of having multiple seeds to allow just one of them to be alive? Perhaps this code should have just waited for ONE of the given nodes to become available, not all of them?

Reply:
> On the other hand, I'm not sure you should have really insisted that ALL seeds be available. Isn't the whole point of having multiple seeds to allow just one of them to be alive? Perhaps this code should have just waited for ONE of the given nodes to become available, not all of them?

If the admin provides a seed which is not a member of the cluster, they made an error and should remove the seed.
If the admin provides a seed which is dead, we cannot proceed with bootstrap anyway, because bootstrap requires all nodes to be alive.
If in the future we will allow bootstrap with some nodes dead, such as with `--ignore-nodes` option, the admin can either remove the ignored nodes from seed list or we can filter them out. But for now we require all nodes.

BTW. if Raft is enabled we already use all seeds explicitly in the discovery algorithm.

Reply:
> The seeds are provided via scylla.yaml that is copied over from somewhere.

All of those old seeds might have left the cluster already. The administrator needs to make sure the list is up-to-date anyway. They need to provide exactly one seed by the way, there's basically no reason to provide multiple seeds today. There will be when we support concurrent boot to quickly boot a multi-node cluster - but that will require Raft.

Note that with `consistent_cluster_management` we already contact all seeds during initial boot. This is required for the correctness of the discovery algorithm. So we basically backport this requirement to non-Raft-enabled boots. But this is `master` branch. So the release which contains this backported requirement will already have Raft enabled by default.

Reply:
Like Avi, I was always under the impression that seed nodes were just a list of "maybe alive nodes" and as soon as you contacted just one of them, you can get from it everything you need - the token ring, gossip, schema, and so on, and never need more than one.

Clearly my impression was incorrect (or over the years, became incorrect?) but I still don't understand why it has to be this way. As soon as *one* of the seed nodes is available, why isn't this enough? I found in some old Datastax documentation the sentence "More than a single seed node per datacenter is recommended for fault tolerance.". But if we have more than one and wait for all of them it doesn't add any fault tolerance.

By the way, I was also surprised about your (unrelated) comment that all the nodes in the cluster must be up when adding a node. Can't we allow one node to be down, when if CL=QUORUM reads succeed? I'm saying this is unrelated to the seed issue, because I thought seeds can be worse than a dead node - I thought they may be long-dead IP addresses that used to be seeds but got removed over time, and things should still work.

Reply:
First of all, let me ask - why are we bikeshedding about this? We can discuss the role of seed nodes elsewhere without wasting time and preventing useful fixes from getting merged, if you feel that it's so important to you.

I can change the code to wait for only one of provided seeds to appear, but it **really doesn't matter**. Really.
I will change it just so we can finish the discussion and do something more productive.

> Clearly my impression was incorrect (or over the years, became incorrect?) but I still don't understand why it has to be this way. As soon as one of the seed nodes is available, why isn't this enough? I found in some old Datastax documentation the sentence "More than a single seed node per datacenter is recommended for fault tolerance.". But if we have more than one and wait for all of them it doesn't add any fault tolerance.

Datastax documentation does not apply. At some point Scylla stopped using seeds during everyday gossiping and started using them only as contact points during initial boot to discover the existing cluster.

There is no reason to provide an invalid initial contact point during boot.

Furthermore, `consistent_cluster_management: True` we require contacting all seeds during the discovery algorithm so we can correctly build the transitive closure of all peers and consistently select the leader who will create initial group 0. So **the requirement is already there**.

> By the way, I was also surprised about your (unrelated) comment that all the nodes in the cluster must be up when adding a node. Can't we allow one node to be down, when if CL=QUORUM reads succeed? I'm saying this is unrelated to the seed issue, because I thought seeds can be worse than a dead node - I thought they may be long-dead IP addresses that used to be seeds but got removed over time, and things should still work.

You're using the old interpretation of seeds, today they mean "initial contact points for boot". Maybe the mistake was that we didn't introduce a completely new configuration parameter for that and instead reused the old "seed" name. In general every node you boot may use a completely different seed, because you take an IP of one of the currently existing nodes in the cluster.

Regarding the requirement of all nodes being UP, some data may be present on only one of the nodes hence we require streaming from all nodes. The user might be doing CL=ONE writes or reads, and we should not lose this data unless the user explicitly tells us they don't care (by providing an unsafe option such as `--ignore-dead-nodes`).

Reply:
> I can change the code to wait for only one of provided seeds to appear, but it really doesn't matter. Really.
>  I will change it just so we can finish the discussion and do something more productive.

Hmm actually, waiting for all seeds **is** better, as explained here: https://github.com/scylladb/scylladb/pull/14100#issuecomment-1570272874
So I will keep defending this solution.

Reply:
> > The seeds are provided via scylla.yaml that is copied over from somewhere.
> 
> All of those old seeds might have left the cluster already. The administrator needs to make sure the list is up-to-date anyway. They need to provide exactly one seed by the way, there's basically no reason to provide multiple seeds today. There will be when we support concurrent boot to quickly boot a multi-node cluster - but that will require Raft.
> 
> Note that with `consistent_cluster_management` we already contact all seeds during initial boot. This is required for the correctness of the discovery algorithm. So we basically backport this requirement to non-Raft-enabled boots. But this is `master` branch. So the release which contains this backported requirement will already have Raft enabled by default.

You're changing it from "at least one must be members of the cluster and up right now" to "all of them". That's a big change. And is it necessary?

Why are the seeds special in any way, besides providing initial contact points? Let's not give them additional roles.

Reply:
> You're changing it from "at least one must be members of the cluster and up right now" to "all of them". That's a big change. And is it necessary?

First of all, "all of them" is already required in Raft mode. So the only thing that I change - is backporting this requirement to non-Raft mode, which becomes the non-default anyway.

Now let me provide an example to show why "wait for all seeds" makes things simple and more elegant.
Consider the standard setup where we boot two nodes X and Y, and seed lists are:
X: {X},
Y: {X, Y}.

X must boot first, Y second.
With "wait for all seeds to appear", when X boots, it will trivially pass the wait as it observes itself.
Y boots and it will pass the wait once it observes Y.

Now consider "wait for one other member to appear".
We can immediately see that this doesn't work because X would hang forever on that.
So we need to complicate the condition and say "if I'm not the first node, wait for one other member to appear".
And now we need to use some logic to decide that "I'm the first node".
We already have `is_first_node()` which looks at the seed list and checks that it's the smallest IP in the seed list. We could use that but is it really simpler?

> Why are the seeds special in any way, besides providing initial contact points? Let's not give them additional roles.

I'm literally using them only as initial contact points, meaning that you need to contact them when you initialize.

Reply:
@kbr-scylla please don't use the phrase "bike shedding". You decided to change the role of seed nodes ("all of them must be up" instead of "one must be up") which has no relation to your patch, and Avi and I are trying to understand why - because it will be easy for you to just wait for any seed node .  Note that it's not about simplicity, it's about correctness. Also try to tell me - what's the point of having more than one seed nodes if you need all of them to be up??? With your plan, it sounds like everyone should have exactly one seed node, because having twe just gives you more trouble for no gain - you now need both of them to be up.

I'm curious, by the way, why Raft mode also needs to all seed nodes to be up and not just one. Maybe you also made a mistake there?

Your example with X waiting for X just means that if you have only yourself as a seed, you don't wait. That's it. A simple if(). You don't need to have the 17th refuse to join the cluster because one of the 3 seed nodes listed are down just to avoid this if().

Reply:
> Note that it's not about simplicity, it's about correctness.

But you never provided what's incorrect about waiting for all seed nodes.
Will it lead to cluster split-brain or something? Intuitively synchronizing with fewer nodes is less safe than synchronizing with more nodes.

> Also try to tell me - what's the point of having more than one seed nodes if you need all of them to be up?

When you boot nodes sequentially, there is no point. You can do: `X: {}, Y: {X}, Z: {Y}, ...`.
There was never a point of providing more than one seed node after we removed seed nodes from every-day gossiping and reduced them to purely initial contact points for boot.

Having more than one seed nodes will be useful for concurrent cluster boot once we support it (in Raft-topology mode). You will be able to configure your seeds like this (for example): `X: {X, Y, Z}, Y: {X, Y, Z}, Z: {X, Y, Z}` and start all of them, they will choose an order of boot among themselves and sequence the ops internally.

> With your plan, it sounds like everyone should have exactly one seed node, because having two just gives you more trouble for no gain - you now need both of them to be up.

And that's true. You already need only one node for doing sequential bootstraps.
Multiple seed nodes will only become useful once we support concurrent bootstraps, as explained above.

It's fine for you to provide multiple seed nodes though, just don't provide dead/left nodes - it won't work (in Raft mode it already won't).

> I'm curious, by the way, why Raft mode also needs to all seed nodes to be up and not just one. Maybe you also made a mistake there?

No, it was not a mistake, it was a conscious decision when we designed the initial cluster discovery algorithm with Kostja. It's required to ensure that we consistently select the initial leader which creates group 0 and then joins everyone else, as I mentioned in an earlier comment.

> Your example with X waiting for X just means that if you have only yourself as a seed, you don't wait. That's it. A simple if(). You don't need to have the 17th refuse to join the cluster because one of the 3 seed nodes listed are down just to avoid this if().

You're proposing to introduce a new definition of "I am the first node (therefore I don't have to wait for others)" into the code. We already have an existing, but different definition of "I am the first node". Maybe let's introduce a third one just for fun?